multiprocess_functions.py

```diff
@@ -0,0 +1,166 @@
+    awk = subprocess.Popen(["awk", "{print $4}"], stdin=grep.stdout, stdout=subprocess.PIPE)  
+    l = awk.communicate()[0].decode('utf-8').split(',')[0].split('-')
+    l = [int(x) for x in l]
+    cores_in_numa = min(cores_in_numa, l[1] - l[0] + 1)
+    core_min_max.append(l)
+
+  core_list = []
+  if cores_per_process <= cores_in_numa:       # Normal case
+    for i in range(numa_nodes):
+      for j in range(cores_in_numa//cores_per_process):
+        core_list = core_list + list(range(core_min_max[i][0] + j*cores_per_process, core_min_max[i][0] + (j+1)*cores_per_process))
+  else:        # single process case or single socket case
+    core_list = range(os.cpu_count()//2)
+
+  return core_list, numa_nodes
+
+def multiprocessing_run(files, max_processes, bash_subprocess):
+  size_dict = dict()
+  for file in files:
+    size_dict[file] = get_file_size(file)
+
+  sorted_size_dict = dict(sorted(size_dict.items(), key=lambda item: item[1], reverse=True))
+
+  total_cores = get_total_cores()
+  cores_per_process = total_cores // max_processes
+  pool = mp.Pool(processes=max_processes)
+
+  queue = [i for i in range(max_processes)]
+  core_list, numa_nodes = get_core_list(cores_per_process)
+  error_files = []
+  def update_queue(result):
+    print(result)
+    index = core_list.index(result[3][0])
+    queue.append(index // cores_per_process)
+    # queue.append(result[3][0] // cores_per_process)
+    if (result[0] != 0):
+      error_files.append(result[1])
+
+  # Iterate over the files and start a new subprocess for each file.
+  print(len(sorted_size_dict))
+  results = [None] * len(sorted_size_dict)
+
+  i = 0
+  for file, value in sorted_size_dict.items():
+    file_path = file
+    process_num = queue.pop(0)
+
+    if max_processes < numa_nodes:
+      if max_processes == 1:
+        if numa_nodes > 1:
+          mem = '0-{}'.format(numa_nodes-1)
+        else:
+          mem = '0'
+      else:
+        mem = '{}-{}'.format(str(process_num * (numa_nodes//max_processes)), str(((process_num + 1) * (numa_nodes//max_processes)) - 1))
+    else:
+      mem = str(process_num//(max_processes//numa_nodes))
+    
+    results[i] = pool.apply_async(bash_subprocess, args=(file_path, mem, core_list[process_num*cores_per_process: (process_num+1)*cores_per_process]), callback = update_queue)
+    i += 1
+    while len(queue) == 0 and i < len(sorted_size_dict):
+        time.sleep(0.05)
+  pool.close()
+  pool.join()
+
+  return error_files
+
+def create_process_list(files, MIN_MEM_PER_PROCESS, MIN_CORES_PER_PROCESS, LOAD_BALANCE_FACTOR):
+  total_cores = get_total_cores()
+  #numa_nodes
+  lscpu = subprocess.Popen(["lscpu"], stdout=subprocess.PIPE)
+  grep = subprocess.Popen(["grep", "NUMA node(s):"], stdin=lscpu.stdout, stdout=subprocess.PIPE)
+  awk = subprocess.Popen(["awk", "{print $3}"], stdin=grep.stdout, stdout=subprocess.PIPE)
+  #Get the output
+  numa_nodes = int(awk.communicate()[0])
```

Comment:
The above block seems repeated from lines 30-35.

Reply:
created a new function get_numa_nodes()